### PR TITLE
SCUMM: Check for and flag demos correctly (PR#931 Fixup)

### PR DESCRIPTION
**Pull request by digitall**
Rewrote the detection matching to use the extra field like the other
two.

This requires extensive testing due to touching code shared for a lot of
games and being sensitive to individual versions of games.

### CODE REVIEW

File: engines/scumm/detection.cpp [hunk 1]
```diff
@@ -489,42 +489,45 @@ static void computeGameSettingsFromMD5(const Common::FSList &fslist, const GameF
...
-				if (md5Entry->platform != Common::kPlatformUnknown) {
-					dr.game.platform = md5Entry->platform;
-				} else if (gfp->platform != Common::kPlatformUnknown) {
-					dr.game.platform = gfp->platform;
-				}
+		if (g->gameid[0] != 0 && scumm_stricmp(md5Entry->gameid, g->gameid)) continue;
+		// The gameid either matches, or is empty. The latter indicates
+		// a generic entry, currently used for some generic HE settings.
+
+		if (g->variant != 0 && !!scumm_stricmp(md5Entry->variant, g->variant)) continue;
```
Comment (bluegr) [inline engines/scumm/detection.cpp]: The two not operators here (!!) are superfluous, and unneeded

File: engines/scumm/detection.cpp [hunk 2]
```diff
@@ -489,42 +489,45 @@ static void computeGameSettingsFromMD5(const Common::FSList &fslist, const GameF
...
+		// Variant matches too
+
+		if (g->preferredTag !=0 && !!scumm_stricmp(md5Entry->extra, g->preferredTag)) continue;
```
Comment (bluegr) [inline engines/scumm/detection.cpp]: Same here, with !!

File: engines/scumm/detection.cpp [hunk 3]
```diff
@@ -489,42 +489,45 @@ static void computeGameSettingsFromMD5(const Common::FSList &fslist, const GameF
...
+		//OH, and do check the extra field in the entry too, otherwise we miss demos
```
Comment (bluegr) [inline engines/scumm/detection.cpp]: Nitpicking: this comment needs formatting to be the same as the rest

File: engines/scumm/detection.cpp (hunk 2 above)
Comment (bluegr) [reply engines/scumm/detection.cpp]: Also, this line is the only change to the detector code in this PR :)

**Comment by digitall** [comment]
@bluegr : I have addressed all your comments and rebased the commit in this PR onto the latest master commit.

One point, is that the "!!" is not just spurious, but seems to be a practice to avoid compiler warnings and sanitize the value to a boolean. See:
http://stupefydeveloper.blogspot.co.uk/2009/09/c-double-exclamation.html

I am not sure, but I think this is not necessary here unless @henke37 did it for some specific reason?

**Comment by bluegr** [comment]
Thanks!

Yeah, the usage of !! does help sanitize a return value to a boolean, but it's not really needed here, as the condition above these does not use it.

Have you tried these changes? Are all the SCUMM games detected properly?

As mentioned, the only real change to the detector is the additional check for the preferredTag field

**Comment by digitall** [comment]
@bluegr No problem. No, I haven't tried these changes yet, beyond checking they compile without warnings.

I can try moving my .scummvmrc file off, and doing a Mass Add with and without this patch to see what difference it makes, but I don't think I have most of the SCUMM game demos so this may be pointless. I suggest that a SCUMM engine dev who has all of the relevant demos tests this and sees if the detection is better as they are better placed to know if this is correct. Not sure who that should be? @sev- ?

**Comment by sev-** [comment]
OK, merging this. Will do some detection testing after the merge.

**Pull request merged**